### PR TITLE
APO zero and APO down loss

### PR DESCRIPTION
**Pull request by Athe-kunal**
In huggingface SmolLM3 [blog post](https://huggingface.co/blog/smollm3), they used APO instead of DPO. This PR implements these algorithms, and I hope that we can include these in the upcoming iterations of the book

### CODE REVIEW

**Comment by Athe-kunal** [comment]
@natolambert 
Let me know about the PR

**Comment by natolambert** [comment]
@Athe-kunal thanks for the PR, see the comment above, also share a reference link on wandb or equivalent showing this working with a reference config.

**Comment by Athe-kunal** [comment]
@natolambert thanks for the response
I responded to the earlier comments

APO Zero and APO Down workspace: [Workspace](https://wandb.ai/ad-finance/RLHF-Book?nw=nwuserathe_kunal)

Let me know if this looks good

**Comment by natolambert** [comment]
Link is private @Athe-kunal

**Comment by Athe-kunal** [comment]
Sorry about that @natolambert 

It was committed to my private team, so I had to make a report to share it

[Report](https://wandb.ai/ad-finance/RLHF-Book/reports/APO-Zero-and-APO-Down-training-runs--VmlldzoxNjU0NTE1Mg?accessToken=w3384zr15ruoa49bhdnuy70dvyv7hsl1bg2z8ofvi4qr9u81s8n5draily1f6abo)

**Comment by natolambert** [comment]
These look good, I'll replicate on my DGX Spark and clean up and merge!

**Comment by Athe-kunal** [comment]
Thanks @natolambert, let me know how it goes on your end

Also, I found one issue with torch < 2.6.0, where torch.load was failing because of vulnerability issues, and it requires `weights_only=True`, so I had to bump the torch version

**Comment by natolambert** [comment]
Ran full reference jobs on the DGX Spark:

- APO Zero: https://wandb.ai/natolambert/rlhf-book/runs/3m2uiyw3
- APO Down: https://wandb.ai/natolambert/rlhf-book/runs/tqzkmyi6

Both look directionally correct. APO Down is still running, but the shape looks right so far. I can help clean up the PR later, but the implementation looks correct.

**Comment by Athe-kunal** [comment]
That's great @natolambert 
Looking forward to the course, and thanks for these resources to learn RLHF

**Comment by natolambert** [comment]
NOTE: THIS IS WRITTEN BY CLAUDE, PR IS NEARLY THERE, I'M MAKING ANOTHER PR TO ADD RUFF CI

Thanks for the PR @Athe-kunal, implementation looks correct and the reference runs check out. A few things before merging:

**1. Docs needed:** Please add APO Zero and APO Down to the algorithm table in code/README.md with:
- Paper link: https://arxiv.org/abs/2408.06266
- Reference run links:
  - APO Zero: https://wandb.ai/natolambert/rlhf-book/runs/3m2uiyw3
  - APO Down: https://wandb.ai/natolambert/rlhf-book/runs/tqzkmyi6

Also add a CHANGELOG.md entry under Unreleased.

**2. pyproject.toml:** The torch>=2.6.0 bump makes sense and we will keep it. However, please revert these unrelated changes:
- requires-python >=3.12,<3.13 -- this is too restrictive, keep it as >=3.12
- pytorch-cu121 to pytorch-cu126 -- unrelated CUDA wheel change, keep that separate (right?)
- Removed comments -- prefer to keep the existing comments in toml file

**Comment by Athe-kunal** [comment]
Hi @natolambert 

Torch 2.6.0 does not have support for cu121, so I had to change it to cu126

```
× No solution found when resolving dependencies for split (markers: platform_machine == 'x86_64' and sys_platform == 'linux'): ╰─▶ Because only torch{platform_machine == 'x86_64' and sys_platform == 'linux'}<=2.5.1+cu121 is available and your project depends on torch{platform_machine == 'x86_64' and sys_platform == 'linux'}>=2.6.0, we can conclude that your project's requirements are unsatisfiable. And because your project requires rlhf-book-code[dev], we can conclude that your project's requirements are unsatisfiable.
```
